src/streams.ts

```diff
@@ -50,25 +54,171 @@ export function compose(streams: NodeJS.ReadWriteStream[]) {
+    if (this._closed) {
+      throw new Error('Wrote to closed writable iterable');
+    }
+    return this._write({value, done: false});
+  }
+
+  /**
+   * Close the queue, indicating that no more values will be written.
+   *
+   * If this method is not called, a consumer iterating over the values will
+   * wait forever.
+   *
+   * The returned promise resolves once the consumer has been notified of the
+   * end of the queue.
+   */
+  async close() {
+    this._closed = true;
+    return this._write({done: true} as any);
+  }
+
+  private async _write(value: IteratorResult<V>) {
+    if (this.blockedOn) {
+      this.blockedOn.resolve(value);
+      this.blockedOn = undefined;
+    } else {
+      const deferred = new Deferred<void>();
+      this.backlog.push({value, deferred});
+      await deferred.promise;
+    }
+  }
+
+  /**
+   * Iterate over values in the queue. Not intended for multiple readers.
+   * In the case where there are multiple readers, some values may be received
+   * by multiple readers, but all values will be seen by at least one reader.
+   */
+  async * [Symbol.asyncIterator](): AsyncIterator<V> {
+    while (true) {
+      let value;
+      const maybeValue = this.backlog.shift();
+      if (maybeValue) {
+        maybeValue.deferred.resolve(undefined);
+        value = maybeValue.value;
+      } else {
+        this.blockedOn = new Deferred();
+        value = await this.blockedOn.promise;
+      }
+      if (value.done) {
+        return;
+      } else {
+        yield value.value;
+      }
+    }
+  }
+}
+
+/**
+ * Implements `stream.Transform` via standard async iteration.
+ *
+ * The main advantage over implementing stream.Transform itself is that correct
+ * error handling is built in and easy to get right, simply by using async/await
+ * and
+ */
+export abstract class AsyncTransformStream<In, Out> extends Transform {
+  private readonly _inputs = new AsyncQueue<In>();
+
+  /**
+   * Implement this method!
+   *
+   * Read from the given iterator to consume input, yield values to write
+   * chunks of your own. You may yield any number of values for each input.
+   *
+   * Note: currently you *must* completely consume `inputs` and return for this
+   *   stream to close.
+   */
+  protected abstract _transformIter(inputs: AsyncIterable<In>):
+      AsyncIterable<Out>;
+
+  private _initialized = false;
+  private _writingFinished = new Deferred<void>();
+  private _initializeOnce() {
+    if (this._initialized === false) {
+      this._initialized = true;
+      const transformDonePromise = (async() => {
+        for
+          await(const value of this._transformIter(this._inputs)) {
+            // TODO(rictic): if `this.push` returns false, should we wait until
+            //     we get a drain event to keep iterating?
+            this.push(value);
+          }
+      })();
+      // TODO(rictic): blindly drain the rest of the inputs if _transformIter
+      //     returns early?
```

Comment:
Maybe we provide an explicit method to drain remaining inputs and throw an error/reject the promise if the inputs are not empty (i.e. if its closed?) as a guard against accidental early exits?  Also, why don't expose closed state on the AsyncQueue even as a readonly public getter?

```typescript
if (!this._inputs.closed) {
  this._writingFinished.reject(new Error('Unfinished inputs queue.'));
} else {
  this._writingFinished.resolve(undefined);
}
```